### PR TITLE
Implement Bloom model for CourseChat

### PR DESCRIPTION
**Pull request by FPeng2288**
Bloom-bot's functionality has been implemented to mimic a Socratic dialogue with help, exit, and restart features. the AI assistant can make associations and conversations based on the context.

### CODE REVIEW

File: src/message/message.service.ts [hunk 1]
```diff
@@ -77,4 +77,20 @@ export class MessageService {
     const message = await this.findOne(messageId); // throws NotFoundException if not found
     await this.messageRepository.remove(message);
   }
+
+  async getContextForChat(chatId: number): Promise<string[]> {
```
Comment (madooei) [inline src/message/message.service.ts]: Why did you add this function? The `findAll` method gives you all the messages for a given `chatId`

File: src/message/message.service.ts [hunk 2]
```diff
@@ -77,4 +77,20 @@ export class MessageService {
...
+    const messages = await this.messageRepository.find({
+      where: { chat: { id: chatId } },
+      order: { createdAt: 'ASC' }, // *order by creation date
+    });
+
+    // *return an array of message content
+    return messages.map((message) => message.content);
+  }
+
+  async clearChatHistory(chatId: number): Promise<void> {
```
Comment (madooei) [inline src/message/message.service.ts]: Why did you add this function? First off, we don't want to delete chat history! Besides, if we want to do that, you can call the `delete` method in the `ChatService` to delete a chat, and that will cascade deleting the messages associated with it. But above all, why do you need this operation?

File: src/ai-services/base-assistant.service.ts [hunk 3]
```diff
@@ -3,9 +3,13 @@ export abstract class BaseAssistantService {
   abstract readonly description: string;
   // Define other shared properties and methods here
 
-  // Shared Method Example
-  public async generateResponse(input: string): Promise<string> {
-    // Implement a shared way to generate responses here, can be overridden by extending classes
-    return `Response from ${this.modelName}: ${input}`;
+  // Modified Method
+  public async generateResponse(
```
Comment (madooei) [inline src/ai-services/base-assistant.service.ts:5]: Why did you change the signature of an existing public function?! If I merge this, many other app parts might break down because other parts assume this operation takes only two arguments! The getResponse should only get a message, and optionally a `chatId`, and then use that `chatId` to retrieve the chat history if it wants.

File: src/conversation/conversation.service.ts [hunk 4]
```diff
@@ -110,16 +111,31 @@ export class ConversationService {
 
     const { message } = continueConversationDto;
 
+    // *Check if the user wants to restart the conversation
+    if (message.trim() === 'restart') {
```
Comment (madooei) [inline src/conversation/conversation.service.ts]: I didn't know this feature existed in Bloom, but if it does, we don't want it! It does not make any sense to "reset" a conversation!!

**Review by madooei — CHANGES_REQUESTED** [review]
Where is the "thought chain"? This is not how bloom works! Where is the part that the AI thinks about the input?

File: src/conversation/conversation.service.ts (hunk 4 above)
Comment (FPeng2288) [reply src/conversation/conversation.service.ts]: They mentioned in the welcome_message "If I'm off track, just say so! If you'd like to reset our dialogue, use the /restart command.", since they are designed to be on Discord, a reset button should be present, I'm just trying to mimic if a similar action can be implemented. So I added this feature, and if this feature is not necessary, then I can remove this feature.

File: src/message/message.service.ts (hunk 2 above)
Comment (FPeng2288) [reply src/message/message.service.ts]: This is just based on a mockup of Bloom's restart feature, because I realized that users may want to start a new conversation within an hour but don't want to be connected to the previous one, so I followed the hints in the welcome message and designed this restart feature. However, I did notice that you mentioned in Notion that "to ensure transparency and maintain the integrity of the educational process, students are not permitted to delete or modify their conversations with the AI", so this is just a test version and I will remove it now.

**Review by madooei — APPROVED** [review]
Looks good @FPeng2288 ! Thank you!
